src/Build/BackEnd/Node/MSBuildClient.cs

```diff
@@ -0,0 +1,661 @@
+                                {
+                                    CommunicationsUtilities.Trace("Server disconnected abruptly");
+                                }
+                                else
+                                {
+                                    CommunicationsUtilities.Trace("Incomplete header read from server.  {0} of {1} bytes read", bytesRead, headerByte.Length);
+                                }
+
+                                _packetPumpShutdownEvent.Set();
+                                continueReading = false;
+                                break;
+                            }
+
+                            NodePacketType packetType = (NodePacketType)Enum.ToObject(typeof(NodePacketType), headerByte[0]);
+
+                            try
+                            {
+                                _packetFactory.DeserializeAndRoutePacket(0, packetType, BinaryTranslator.GetReadTranslator(localPipe, _sharedReadBuffer));
+                            }
+                            catch (Exception e)
+                            {
+                                // Error while deserializing or handling packet.  Abort.
+                                CommunicationsUtilities.Trace("Packet factory failed to recieve package. Exception while deserializing packet {0}: {1}", packetType, e);
+                                ExceptionHandling.DumpExceptionToFile(e);
+
+                                _packetPumpShutdownEvent.Set();
+                                continueReading = false;
+                                break;
+                            }
+
+                            // Start reading the next package header.
+#if FEATURE_APM
+                            result = localPipe.BeginRead(headerByte, 0, headerByte.Length, null, null);
+#else
+                            readTask = CommunicationsUtilities.ReadAsync(localPipe, headerByte, headerByte.Length);
+#endif
+                        }
+
+                        break;
+
+                    case 1:
+                        // Fulfill the request for termination of the message pump.
+                        CommunicationsUtilities.Trace("Terminate message pump thread.");
+                        continueReading = false;
+                        break;
+
+                    default:
+                        // Ignore unknown package.
+                        ErrorUtilities.ThrowInternalError("waitId {0} out of range.", waitId);
+                        // TODO: cover this case with ETW.
+                        break;
+                }
+            }
+            while (continueReading);
+
+            CommunicationsUtilities.Trace("Ending read loop");
+        }
+#endregion
+
+
+        /// <summary>
+        /// Connects to MSBuild server.
+        /// </summary>
+        /// <returns> Whether the client connected to MSBuild server successfully.</returns>
+        private bool ConnectToServer(int timeout)
```

Comment:
nit: There is a convention that methods returning `bool` and not throwing exceptions are named with `Try` prefix.